### ISSUE TITLE
Question: could `tags` ensure every steps running on specific tagged worker?

### ISSUE THREAD

**Issue by SimonXming** [concourse/concourse#3768]
## Bug Report

We want to make sure pipeline only running on dedicated worker pool. We already have several workers with specific `tag`(for example `TEST`). Also, we have a pipeline adding the `tags` attribute (`tags: [TEST]`) for `resource_types`, `resources`, `jobs.task`, `jobs.put`.

## Steps to Reproduce

Example: https://gist.github.com/SimonXming/c5b475daaf6781feb3bb113fde1dd18a

## Expected Results

All pipeline steps(including: `get`, `put`, `check`, `task`) should be running on workers with `TEST` tag.

## Actual Results

Some `get` steps will be running on untagged workers.

```
fly -t dev containers
2662673f-9eaf-4a40-701f-02e5d6dcfedb  untagged-worker  pipeline-test-tags         job-1                 4     219294  get    small-file          n/a
d133dca7-d371-41a1-541b-b3d42b6b3978  worker-with-tags pipeline-test-tags         job-1                 4     219294  task   sleep-job           n/a
eb08a071-6cc4-46c2-6018-ee84ee6bab46  untagged-worker  pipeline-test-tags         job-1                 4     219294  get    simple-timer        n/a
```

## Version Info

* Concourse version: 5.0.1
* Deployment type (BOSH/Docker/binary): Docker
* Infrastructure/IaaS: Kubernetes
* Browser (if applicable): Chrome
* Did this used to work? Not

**Comment by vito** [comment]
`tags:` on a resource definition only influences where the `check` for the resource happens. It does not propagate to `get` or `put` - it has to be specified there too. Quite painful and verbose, but it is what it is.

**Comment by vito** [comment]
We're thinking about actually implementing something like this soon though; it came up in discussions regarding the resource checking queue we're implementing (#3788).

The semantics of resources with `tags:` is a little awkward to reason about as it's hard to tell if the tags are there purely to influence placement or if the tags are there to actually affect the semantics of the resource, i.e. it literally has to run in `vsphere` or on a worker with ECR access in order to function. In the latter case it makes more sense for the `tags:` to "stick" to the resource throughout `get` and `put` steps and even going down to the versions associated to it.

I'll re-open this for now, but I'll probably re-frame it as an enhancement and close it once the new issue is open. 🙂 Thanks for your feedback, it helps to see that that's how folks expected it to work in the first place.

**Comment by SimonXming** [comment]
Cool! I'm really looking forward to seeing this feature. One small suggestion though, could we also provide a syntax sugar for add `tags` at the pipeline level?

**Comment by vito** [comment]
@SimonXming There's an open issue for that: https://github.com/concourse/concourse/issues/371 - but we won't be doing it at the same time as we'll be focusing more on the resource semantics aspect.

**Comment by SimonXming** [comment]
> @SimonXming There's an open issue for that: #371 - but we won't be doing it at the same time as we'll be focusing more on the resource semantics aspect.

Agreed. Is there any open issue about the progress of resource semantics?

**Comment by vito** [comment]
@SimonXming No issue yet - just a card: https://github.com/orgs/concourse/projects/31#card-21636186

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
